### DISCUSSION TITLE
cy.intercept, beforeEach, and inconsistent wait()

### DISCUSSION THREAD

**Discussion by bared-dave** [cypress-io/cypress#16842] [CI setup]
Hi there,

I'm working on introducing Cypress tests into our codebase, and am having inconsistent results when trying to intercept API requests in beforeEach block, and inconsistent results in `visit()` waiting for pages to load.

I've created a minimal repo which shows the problem. https://github.com/bared-dave/cypress-remote-url-test/

Sometimes this works locally, sometimes it doesn't. It almost never works in CI context. https://github.com/bared-dave/cypress-remote-url-test/actions

If we take this test as an example:

```
    beforeEach(() => {
      cy.setLocalStorage('headless:subscribe', { value: 'hidden', expiry: 0 })
      cy.intercept({ url: 'https://ipapi.co/json/?*', middleware: true }, mockLocationAPI('au')).as('LocationAU')
      cy.intercept({ url: '**/api/*/graphql.json', middleware: true }, mockShopifyAPI)
      cy.visit('/products/aluminium-black-lace-ups')
      cy.wait(['@GetShopSettings', '@CheckoutCreate', '@GetProduct', '@LocationAU'])
    })

    it('clicking trigger should display available currencies', () => {
      cy.get(priceNormalSelector).first().should('have.text', '$289')
      cy.get(priceDiscountedSelector).first().should('have.text', '$210')
      cy.get(triggerSelector).click({ force: true })
      cy.get(audSelector).should('be.visible')
      cy.get(nzdSelector).should('be.visible')
      cy.get(usdSelector).should('be.visible')
      cy.get(usdSelector).click({ force: true })
      cy.wait(['@GetShopSettings', '@CheckoutCreate'])
      cy.get(priceNormalSelector).first().should('have.text', 'USD$246')
      cy.get(priceDiscountedSelector).first().should('have.text', 'USD$179')
      cy.get(nzdSelector).click({ force: true })
      cy.wait(['@GetShopSettings', '@CheckoutCreate'])
      cy.get(priceNormalSelector).first().should('have.text', 'NZD$315')
      cy.get(priceDiscountedSelector).first().should('have.text', 'NZD$229')
      cy.get(audSelector).click({ force: true })
      cy.wait(['@GetShopSettings', '@CheckoutCreate'])
      cy.get(priceNormalSelector).first().should('have.text', '$289')
      cy.get(priceDiscountedSelector).first().should('have.text', '$210')
    })
```

I have noticed that sometimes when `visit()` is called, the XHR requests that I'm attempting to intercept (and replace with a fixture) are called **before** the `wait()` function is called, so then the `wait()` function times out and the test fails because the XHR request has already been and gone.

If I remove the `wait()`, sometimes the tests pass locally if the XHRs are intercepted before the individual tests are run, but most of the time they don't. I've been banging my head against the wall for 2 days trying to fix this.

Is there a way to make these tests more consistent that I'm missing?

See below screenshot:

<img width="470" alt="Screen Shot 2021-06-08 at 12 40 36 pm" src="https://user-images.githubusercontent.com/82138784/121114375-f0f0fd00-c856-11eb-8aa5-03c0c83aa436.png">

<img width="468" alt="Screen Shot 2021-06-08 at 12 41 25 pm" src="https://user-images.githubusercontent.com/82138784/121114392-f51d1a80-c856-11eb-9899-c545c8c386cd.png">

For reference we are using Gatsby, Gatsby cloud and Gitlab. I have setup Gatsby cloud webhook notification that calls a Gitlab webhook to trigger a job which runs the Cypress tests against a remote URL.

**Comment by bahmutov** [comment]
1. Why is the app throwing so many errors "Cannot set property status ..."?
2. In this test and in others you are creating an alias `LocationNZ` but waiting for `LocationAU`
```js
  context('NZD', () => {
    beforeEach(() => {
      cy.setLocalStorage('headless:subscribe', { value: 'hidden', expiry: 0 })
      cy.intercept({ url: 'https://ipapi.co/json/?*', middleware: true }, mockLocationAPI('nz')).as('LocationNZ')
      cy.intercept({ url: '**/api/*/graphql.json', middleware: true }, mockShopifyAPI)
      cy.visit('/products/aluminium-black-lace-ups')
      cy.wait(['@GetShopSettings', '@CheckoutCreate', '@GetProduct', '@LocationAU'])
    })
```
Fixing the alias names fixes the tests for me

<img width="1138" alt="Screen Shot 2021-06-08 at 8 51 08 AM" src="https://user-images.githubusercontent.com/2212006/121188064-bd05df80-c836-11eb-9acd-1b0ba844d2d6.png">

**Reply by bared-dave** [reply]
@bahmutov I've fixed that, however the tests are still failing periodically. See video from test that I ran locally that failed.

https://user-images.githubusercontent.com/82138784/121270526-3cfa7b00-c905-11eb-9bf4-b8c1bba588ce.mp4

**Reply by jennifer-shehane** [reply]
I see these `google.firestore.v1...` XHRs seem to not be resolving when the test hangs on the cy.wait(). Not sure it it's relevant, but are these requests necessary for you tests? You could try using [`blockHosts` config](https://on.cypress.io/configuration#blockHosts) to block those and see if it helps. 

<img width="1564" alt="Screen Shot 2021-06-11 at 1 05 13 PM" src="https://user-images.githubusercontent.com/1271364/121731241-49720580-cab6-11eb-9131-2412a4396e88.png">
